sled-agent/src/sim/upstairs.rs

```diff
@@ -0,0 +1,164 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Simulated Crucible upstairs implementation
+
+use super::storage::Storage;
+use omicron_common::api::external::Error;
+use omicron_uuid_kinds::SledUuid;
+use propolis_client::VolumeConstructionRequest;
+use slog::Logger;
+use std::collections::HashMap;
+use std::collections::VecDeque;
+use std::net::SocketAddr;
+use std::str::FromStr;
+use std::sync::Arc;
+use std::sync::Mutex;
+use uuid::Uuid;
+
+#[derive(Default)]
+pub(crate) struct SimulatedUpstairsInner {
+    id_to_vcr: HashMap<Uuid, VolumeConstructionRequest>,
+    sled_to_storage: HashMap<SledUuid, Storage>,
+}
+
+#[derive(Clone)]
+pub struct SimulatedUpstairs {
+    log: Logger,
+    inner: Arc<Mutex<SimulatedUpstairsInner>>,
+}
+
+impl SimulatedUpstairs {
+    pub fn new(log: Logger) -> SimulatedUpstairs {
+        SimulatedUpstairs {
+            log,
+            inner: Arc::new(Mutex::new(SimulatedUpstairsInner::default())),
+        }
+    }
+
+    pub(crate) fn register_storage(
+        &self,
+        sled_id: SledUuid,
+        storage: &Storage,
+    ) {
+        let mut inner = self.inner.lock().unwrap();
+        inner.sled_to_storage.insert(sled_id, storage.clone());
+
+        info!(self.log, "registered sled {sled_id} storage");
+    }
+
+    /// Map id to VCR for later lookup
+    pub fn map_id_to_vcr(
+        &self,
+        id: Uuid,
+        volume_construction_request: &VolumeConstructionRequest,
+    ) {
+        let mut inner = self.inner.lock().unwrap();
+        // XXX what if existing entry? assert?
```

Comment:
Can callers "remap" IDs to new VCRs? If not, seems reasonable to either assert or return a `Result`, I think? (I guess a variant of this question would be: is it reasonable for a caller to call this multiple times with the _same_ ID -> VCR mapping, and if so should that be okay?)

Reply:
Actually yeah, VCR changes happen all the time due to read-only parent removal and the various replacements. Updated this comment in da4aca6da.